consensus/impl/dpos/blockfactory_test.go

```diff
@@ -13,17 +13,18 @@ func TestBlockFactory_context(t *testing.T) {
 	factory.initContext()
 
 	timeoutCtx, _ := context.WithTimeout(factory.ctx, time.Millisecond)
+	contextDeadline, _ := timeoutCtx.Deadline()
 	cancelCtx, cancelFunc := context.WithTimeout(factory.ctx, time.Minute)
 	quitCtx, _ := context.WithTimeout(factory.ctx, time.Minute)
-	time.Sleep(time.Millisecond * time.Duration(2))
 
 	// first channel is done by deadline
 	select {
 	case <-timeoutCtx.Done():
 		err := timeoutCtx.Err()
 		assert.Equal(t, context.DeadlineExceeded, err, "err by deadline")
-		//assert.Equal(t, err, context.Cause(timeoutCtx), "cause and err is differ")
-	default:
+		// check if timeout occured not before deadline
+		assert.True(t, time.Now().After(contextDeadline))
+	case <-time.NewTimer(time.Second * 5).C:
 		assert.Fail(t, "deadline expected, but not")
```

Comment:
consider completing this sentence. it is not clear

Reply:
Like below? 
```Context should be done by deadline.```

Reply:
"context ... done"? I don't understand it

Reply:
"deadline expected, but not happen" may be fine

Reply:
The channel is done if its lifecycle is done by timeout or called cancel. And, `timeoutCtx.Done()` will block until the timeoutCtx is done. 

I can't have more ideas to clearly reveal what makes this test in failure. I will appreciate if you give me a good phrase.

Reply:
The one I wrote above is not good?

Reply:
My reply was on RACE CONDITION. 😆